### DISCUSSION TITLE
Multicomponent phase-field simulations

### DISCUSSION THREAD

**Discussion by jerewang20** [idaholab/moose#25451] [Q&A Modules: Phase field]
I am fairly new to phase-field simulations, but I wanted to try using them to explore liquid-liquid phase separations.  I did see some examples for multiphase models, but I am struggling a bit to implement a multi-component model.  I would like to simulate a system with 4 charged species (two positive and two negative) and solvent, and I was wondering what is the best way to ensure the concentrations of all species sum to 1, and ensure that charge neutrality is maintained.

**Comment by dschwen** [comment]
@vprithiv just worked on this! Not with charged liquids though.

**Reply by dschwen** [reply]
One way to ensure the concentrations of all species sum to 1 is to remove one concentration and substitute it by `1-sum(c_i)`. Another way would be to use the Lagrange Multiplier approach we have implemented.

**Reply by roystgnr** [reply]
I'd go farther and just have two variables: `p` and `n`, for the concentrations of one of the positive and one of the negative species.  Then the second positive species has concentration `0.5-p`, the second negative `0.5-n`.  (Assuming all charged species are singly-charged; the same trick works if some are doubly- or whatever but the numbers differ)

**Reply by vprithiv** [reply]
This is what I have been recently working on: simulating a two-phase flow using a combined Navier-Stokes and Cahn-Hilliard equations. I employed a single order parameter, which can take values of -1 or 1. The phase field was primarily used to track the interface between the two fluids. Material properties in the Navier-Stokes equations, namely density and viscosity, were defined as functions of the order parameter. If you are working on anything similar or find this information helpful, I can provide you with the relevant references.

**Reply by jerewang20** [reply]
@dschwen I did see the Lagrange Multiplier kernel, and I'll look into it more!  It might also help with the charge neutrality.  I was able to directly add the electrostatic penalty to the mixing energy for enforce charge neutrality in my simulations, but it does seem to be computationally costly, and so I was wondering if there was a more efficient way to impose constraints on the component concentrations.  

If you have a good example of the lagrange multiplier being implemented somewhere I think that would help me a lot!

**Reply by jerewang20** [reply]
@vprithiv I am interested in implementing hydrodynamics into my phase field simulations, although currently I am more focused on the spinodal decomposition aspect of polyelectrolyte coacervates systems.  Still, I would love to read up more on coupling Navier-Stokes with Cahn-Hilliard!

**Comment by jerewang20** [comment]
> I'd go farther and just have two variables: `p` and `n`, for the concentrations of one of the positive and one of the negative species. Then the second positive species has concentration `0.5-p`, the second negative `0.5-n`. (Assuming all charged species are singly-charged; the same trick works if some are doubly- or whatever but the numbers differ)

I will give it a shot!  I do have a significant amount of solvent in both phases, but perhaps I just need to rewrite the free energies of mixing.

**Reply by vprithiv** [reply]
These may help:

P Yue et al., Phase-field simulations of interfacial dynamics in viscoelastic fluids using finite elements with adaptive meshing
C Zhou et al., 3D phase-field simulations of interfacial dynamics in Newtonian and viscoelastic fluids
